applications/bwa_alignment/template.yaml

```diff
@@ -0,0 +1,173 @@
+          {{- if $.S3Secret }}
+          secret: {{ $.S3Secret }}
+          {{- end }}
+      {{- end }}
+jobs:
+  bwa-mem2-index:
+    cwd: /data
+    name: bwa-mem2-index
+    image:
+      uri: {{.ContainerImage | default "docker://quay.io/biocontainers/bwa-mem2:2.2.1--he513fc3_0"}}
+    mounts:
+      bwa-data:
+        location: /data
+    command:
+      - /bin/bash
+      - '-c'
+      - |
+        if [ -f reference.fa.gz ] && [ ! -f reference.fa ]; then
+          gunzip -f reference.fa.gz
+        fi
+
+        if [ -f reference.fa ] && [ ! -f reference.fa.bwt.2bit.64 ]; then
+          bwa-mem2 index reference.fa
+        elif [ ! -f reference.fa ]; then
+          exit 1
+        fi
+    resource:
+      cpu:
+        cores: {{.IndexCpuCores | default 16}}
+      memory:
+        size: {{.IndexMemory | default 16}}GiB
+    id: job-{{.JobId | default "bwamem2"}}
+  bwa-mem2-align:
+    cwd: /data
+    image:
+      uri: {{.ContainerImage | default "docker://quay.io/biocontainers/bwa-mem2:2.2.1--he513fc3_0"}}
+    mounts:
+      bwa-data:
+        location: /data
+    command:
+      - /bin/bash
+      - '-c'
+      - |
+        if [ -f sample_1.fastq.gz ]; then
+          gunzip -f sample_1.fastq.gz
+        fi
+
+        if [ -f sample_2.fastq.gz ]; then
+          gunzip -f sample_2.fastq.gz
+        fi
+
+        if [ ! -f sample_1.fastq ] || [ ! -f sample_2.fastq ]; then
+          exit 1
+        fi
+
+        # Add read group if provided
+        RG_OPTS=""
+        {{- if .ReadGroupString }}
+        RG_OPTS="-R '{{.ReadGroupString}}'"
+        {{- end }}
+
+        # Get any additional alignment options
+        EXTRA_OPTS="{{.AlignmentExtraOptions | default ""}}"
+
+        # Run BWA-MEM2 alignment
+        bwa-mem2 mem -t {{.AlignThreads | default 16}} ${RG_OPTS} ${EXTRA_OPTS} reference.fa sample_1.fastq sample_2.fastq > aligned.sam
+        
+        echo "Alignment Date: $(date)" > aligned.stats.txt
+        echo "Reference Genome: {{.ReferenceGenomeName | default "GCF_000005845.2_ASM584v2"}}" >> aligned.stats.txt
+        total_reads=$(grep -v "^@" aligned.sam | wc -l)
+        echo "Total reads: $total_reads" >> aligned.stats.txt
+        mapped_reads=$(grep -v "^@" aligned.sam | grep -v "XA:Z:" | wc -l)
+        echo "Mapped reads: $mapped_reads" >> aligned.stats.txt
+        echo "File sizes:" >> aligned.stats.txt
+        du -h aligned.sam >> aligned.stats.txt
+    requires:
+      - bwa-mem2-index
+    resource:
+      cpu:
+        cores: {{.AlignCpuCores | default 16}}
+      memory:
+        size: {{.AlignMemory | default 16}}GiB
+  msamtools-filter:
+    cwd: /data
+    image:
+      uri: {{.MsamtoolsContainerImage | default "docker://quay.io/biocontainers/msamtools:1.1.2--h7132678_0"}}
+    mounts:
+      bwa-data:
+        location: /data
+    command:
+      - /bin/bash
+      - '-c'
+      - |
+        # Test various filter thresholds to find what works
```

Comment:
Including a filtering step is a good idea. May i suggest making it (1) deterministic by allowing the user to provide options for `msamtools filter` and (2) optional by skipping the job if no options are provided. The reason i suggest this is that users might want to treat all their samples the same way.
If you do decide to keep the filtering step this way I would suggest not applying it three times to the full (again potentially very large file) three times. You could pick a random subset of a given size to test the filters with, for example, and then apply the selected filter to the full output.